### ISSUE TITLE
[Question] Help in migrating Keras LSTM to Darts LSTM 

### ISSUE THREAD

**Issue by gsamaras** [unit8co/darts#975]
I have a Keras LSTM with good accuracy but I would like to use Darts instead, however I am having trouble training the model - it seems that the hyperparameters that I used in Darts are not helping the model with learning.

In Keras I do:
```
from keras.models import Sequential
from keras.layers import Bidirectional, Dense, LSTM

n_steps = 3
out = 1
n_features = 1
model = Sequential()
model.add(
    Bidirectional(LSTM(100, activation="relu"), input_shape=(n_steps, n_features))
)
model.add(Dense(out))
model.compile(optimizer="adam", loss="mse")
model.fit(X, y, epochs=200, verbose="auto")
```
and this produced good accuracy. I want to achieve comparable accuracy with the LSTM of Darts; here is one of my attempts:

```
from darts.models import RNNModel

my_model = RNNModel(
    model="LSTM", # Not a Bidirectional LSTM, but I couldn't figure out how to do it
    hidden_dim=20,
    n_rnn_layers=100,
    dropout=0,
    batch_size=64,
    n_epochs=200,
    optimizer_kwargs={"lr": 1e-3},
    log_tensorboard=True,
    random_state=0,
    training_length=24,
    input_chunk_length=3,
    output_chunk_length=1,
  torch_device_str="cuda:0"
)
my_model.fit(train, val_series=val, verbose=True)
```

But the model doesn't really learn. On the same dataset, I get orders of magnitude worse MAPE (as well as other accuracy metrics). Can someone help? I think I don't really understand how to map the hyperparameters. E.g. the `training_length`. I haven't passed Adam optimizer in Darts LSTM (I think it uses it by default), but I would be hugely surprised if this was the problem.

Can you please help me with my Odyssey in creating a system that uses only Darts? :)

My data are univariate and look like this:

![data](https://user-images.githubusercontent.com/12050610/170643498-64a051b7-d519-4ef1-9275-6b76416157cc.png)

---

RNNModel [ref](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.rnn_model.html?highlight=rnnmodel#darts.models.forecasting.rnn_model.RNNModel) and [example](https://unit8co.github.io/darts/examples/04-RNN-examples.html).

**Comment by hrzn** [comment]
Hi, it might be hard to replicate the *exact* same model. But the fact that it doesn't learn anything is suspicious. Did you scale your series? Also `n_rnn_layers=100` seems too much. Have you tried e.g., using a value of 1 first? It's probably a good idea if `input_chunk_length` is somewhat commensurate to your "period" or characteristic time scale of your series - perhaps 3 is too low? You could try a larger value (e.g. 10, 50, 100 or more). `output_chunk_length` is not used in `RNNModel` so you shouldn't set it (it's always 1). `training_length` should be longer than `input_chunk_length`.

**Comment by hrzn** [comment]
OK, I would definitely try scaling the series first.

**Comment by KaramiF** [comment]
> Surprisingly it did make a difference with Darts LSTM, but still MAPE is needs significant improvement. Here is my complete example:
> 
> ```
> import numpy as np
> import pandas as pd
> import matplotlib.pyplot as plt
> 
> from darts import TimeSeries
> from darts.dataprocessing.transformers import Scaler
> from darts.metrics import mae, mape, mse, rmse, r2_score
> from darts.models import RNNModel, ExponentialSmoothing, BlockRNNModel
> from torch import optim
> from typing import Optional, Union
> 
> def display_forecast(
>     pred_series: TimeSeries,
>     observed_series: TimeSeries,
>     forecast_type: str,
>     start_date: Optional[Union[pd.Timestamp, float, int]] = None,
> ) -> None:
>     """
>     Plot observed versus predicted data.
> 
>     :param pred_series (TimeSeries): The predicted series
>     :param observed_series (TimeSeries): The observed series
>     :param forecast_type (str): Horizon, e.g. 7 days
>     :param start_date (Optional[Union[Timestamp, float, int]]): Start of 'observed_series'
>     :return: None
>     """
>     plt.figure(figsize=(8, 5))
>     if start_date:
>         observed_series = observed_series.drop_before(start_date)
>     observed_series.univariate_component(0).plot(label="actual")
>     pred_series.plot(label=("historic " + forecast_type + " forecasts"))
>     plt.title(
>         "MAPE: {}".format(mape(observed_series.univariate_component(0), pred_series))
>     )
>     plt.legend()
> 
>     MSE = mse(observed_series.univariate_component(0), pred_series)
>     RMSE = rmse(observed_series.univariate_component(0), pred_series)
>     MAE = mae(observed_series.univariate_component(0), pred_series)
>     MAPE = mape(observed_series.univariate_component(0), pred_series)
>     R2 = r2_score(observed_series.univariate_component(0), pred_series)
>     # std = np.std(observed_series.univariate_component(0))
> 
>     print(f" MSE: {MSE}\n RMSE: {RMSE}\n MAE: {MAE}\n MAPE: {MAPE}\n R2: {R2}")
> 
> def define_LSTM_model(
>     train_set: Optional[TimeSeries] = None,
>     val_set: Optional[TimeSeries] = None,
>     gridsearch: bool = False,
> ) -> RNNModel:
>     """
>     Setup LSTM model's architecture.
> 
>     :param train_set (Optional[TimeSeries]): Train set (used in grid search)
>     :param val_set (Optional[TimeSeries]): Validation set (used in grid search)
>     :param gridsearch (Optional[bool]): Perform grid search or not
>     :return: An LSTM model
>     """
>     if gridsearch == True:
>         parameters = {
>           #TODO
>         }
>         
>         # Randomized gridsearch
>         res = RNNModel.gridsearch(
>             parameters=parameters,
>             series=train_set,
>             val_series=val_set,
>             start=0.1,  # starting point in training set
>             last_points_only=False,
>             metric=mape,
>             reduction=np.mean,
>             n_jobs=-1,
>             #n_random_samples=0.99,  # % of full search space to evaluate
>             verbose=True,
>         )
>         best_model, dict_bestparams = res
>         print(f"dict_bestparams: {dict_bestparams}")
>         rnn_model = best_model
>     else:
>         rnn_model = RNNModel(
>             model="LSTM",
>             hidden_dim=128, # 1024 gives MAPE 36
>             n_rnn_layers=1,
>             dropout=0,
>             batch_size=64,
>             n_epochs=50,
>             optimizer_cls = optim.Adam,
>             optimizer_kwargs={'lr': 1e-3},
>             lr_scheduler_cls = optim.lr_scheduler.ReduceLROnPlateau,
>             lr_scheduler_kwargs= {"optimizer": optim.Adam, "threshold": 0.0001, "verbose": True},
>             log_tensorboard=True,
>             random_state=0,
>             training_length=4,
>             input_chunk_length=3,
>             torch_device_str="cuda:0",
>             model_name="LSTM",
>         )
>     return rnn_model
> 
>     
> if __name__ == "__main__":
>     series = pd.read_csv("my_data.csv")
> 
>     trainset_size = 0.60
>     train, val = series.split_after(trainset_size)
>     
>     # Normalize the time series (note: we avoid fitting the transformer on the validation set)
>     transformer = Scaler()
>     train_transformed = transformer.fit_transform(train)
>     val_transformed = transformer.transform(val)
>     series_transformed = transformer.transform(series)
> 
>    rnn_model = define_LSTM_model(train_set=train, val_set=val, gridsearch=False)
>    rnn_model.fit(series=train_transformed, val_series=val_transformed, verbose=True)
>         
>     pred_series = rnn_model.historical_forecasts(
>         series_transformed,
>         start=trainset_size,
>         forecast_horizon=1,
>         stride=1,
>         retrain=False,
>         verbose=True,
>     )
>     
>     display_forecast(pred_series, series_transformed, "1 horizon", start_date=trainset_size)
> ```
> 
> ![Untitled](https://user-images.githubusercontent.com/12050610/170657291-d922f9f8-bae2-4b48-aa2a-4460e4b095df.png)
> 
> Still it needs much improved, but it's a start, thanks!

may I ask for your python and darts version? I have a dependency issue  with  version '3.10.4')  which results in an error while I am importing

**Comment by gsamaras** [comment]
I have scrapped that virtual environment, so I don't know, sorry.
